### PR TITLE
Remove old rendering backend.

### PR DESCRIPTION
**Pull request by glennw**
This removes paint threads, rust-layers dependency, and changes
optional webrender types to be required.

The use_webrender option has been removed, however I've left
the "-w" command line option in place for now so that wpt
runner can continue to pass that. Once it's removed from there
we can also remove the -w option.

Once this stage is complete, it should be fine to change the
display list building code to generate webrender display
lists directly and avoid the conversion step.

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/13711)

### CODE REVIEW

File: components/servo/Cargo.lock [hunk 1]
```diff
@@ -2693,6 +2658,11 @@ dependencies = [
 name = "webrender"
 version = "0.6.0"
 source = "git+https://github.com/servo/webrender#b31b4cf76324bfbe82d27a059e1330ef5ae34c35"
+replace = "webrender 0.6.0"
```
Comment (Ms2ger) [inline components/servo/Cargo.lock]: Huh?

**Comment by glennw** [comment]
We should get the stacking context patch from @mrobinson reviewed and merged before this. This can still be reviewed, and I'll rebase on top of that when it lands.

File: components/canvas/canvas_paint_thread.rs [hunk 2]
```diff
@@ -542,14 +542,12 @@ impl<'a> CanvasPaintThread<'a> {
...
-                                           element.into());
-            }
+            let size = self.drawtarget.get_size();
+            self.webrender_api.update_image(self.webrender_image_key,
+                                            size.width as u32,
+                                            size.height as u32,
+                                            webrender_traits::ImageFormat::RGBA8,
+                                            element.into());
 
             let pixel_data = CanvasPixelData {
```
Comment (emilio) [inline components/canvas/canvas_paint_thread.rs]: Please fill a follow-up to remove the `image_data` field from `CanvasPixelData`.

File: components/compositing/compositor.rs [hunk 3]
```diff
@@ -756,18 +614,7 @@ impl<Window: WindowMethods> IOCompositor<Window> {
             }
 
             (Msg::CollectMemoryReports(reports_chan), ShutdownState::NotShuttingDown) => {
```
Comment (emilio) [inline components/compositing/compositor.rs]: At this point we could also remove this message entirely. Please do it or file a followup for it.

File: components/compositing/compositor.rs [hunk 4]
```diff
@@ -1213,65 +765,11 @@ impl<Window: WindowMethods> IOCompositor<Window> {
...
-                                point: Point2D<f32>) {
-        if self.move_layer(pipeline_id, layer_id, TypedPoint2D::from_untyped(&point)) {
-            self.perform_updates_after_scroll();
-            self.send_viewport_rects_for_all_layers()
-        } else {
-            self.fragment_point = Some(point)
-        }
+                                _pipeline_id: PipelineId,
+                                _layer_id: LayerId,
+                                _point: Point2D<f32>) {
```
Comment (emilio) [inline components/compositing/compositor.rs]: So if I understand this code correctly, this is only called on resize to keep the scroll where the fragment of the URL points to. The code in `script_thread.rs` has a `FIXME(pcwalton)` saying it's very bogus. I think it's a regression we could take, what do you think?

**Review by emilio — COMMENTED** [review]
Only a few comments, not an in-depth review.

File: components/compositing/compositor.rs [hunk 5]
```diff
@@ -229,24 +200,14 @@ pub struct IOCompositor<Window: WindowMethods> {
...
-    last_mouse_move_recipient: Option<PipelineId>,
-
     /// Whether a scroll is in progress; i.e. whether the user's fingers are down.
     scroll_in_progress: bool,
 
     /// The webrender renderer, if enabled.
-    webrender: Option<webrender::Renderer>,
+    webrender: webrender::Renderer,
 
     /// The webrender interface, if enabled.
```
Comment (glennw) [inline components/compositing/compositor.rs]: Fixed

File: components/compositing/compositor.rs [hunk 6]
```diff
@@ -358,26 +306,9 @@ fn initialize_png(width: usize, height: usize) -> RenderTargetInfo {
 
     gl::bind_texture(gl::TEXTURE_2D, 0);
 
-    let renderbuffer_ids = if opts::get().use_webrender {
```
Comment (glennw) [inline components/compositing/compositor.rs]: That piece of code was WR-only, but it's now not required by WR at all - so I've removed it completely.

File: components/compositing/compositor_thread.rs [hunk 7]
```diff
@@ -8,16 +8,14 @@ use SendableFrameTree;
 use compositor::CompositingReason;
 use euclid::point::Point2D;
 use euclid::size::Size2D;
-use gfx_traits::{Epoch, FrameTreeId, LayerId, LayerProperties, PaintListener};
+use gfx_traits::{LayerId};
```
Comment (glennw) [inline components/compositing/compositor_thread.rs]: Fixed

File: components/canvas/canvas_paint_thread.rs (hunk 2 above)
Comment (glennw) [inline components/canvas/canvas_paint_thread.rs]: https://github.com/servo/servo/issues/13734

File: components/compositing/compositor.rs (hunk 3 above)
Comment (glennw) [inline components/compositing/compositor.rs]: https://github.com/servo/servo/issues/13735

File: components/compositing/compositor.rs (hunk 4 above)
Comment (glennw) [inline components/compositing/compositor.rs]: It's also used for scrolling to a # anchor in a page. However, it doesn't work with webrender right now anyway, so it's not technically a regression. https://github.com/servo/servo/issues/13736

**Comment by glennw** [comment]
This passes bors try run now, and I've addressed the comments so far. r? @larsbergstrom @metajack @pcwalton and anyone else who has time to look over some of these changes.

**Comment by larsbergstrom** [comment]
Reviewed 51 of 53 files at r1, 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[components/compositing/compositor.rs, line 771 at r1](https://reviewable.io:443/reviews/servo/servo/13711#-KTrTZZr0Z4fhEPNXpKG-r1-771:-KTvcIsmSbEdQxD_msgs:b-86xeb2) ([raw file](https://github.com/servo/servo/blob/c233541c13bf792afc2691d4df745bd1f5507a33/components/compositing/compositor.rs#L771)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary>
> It's also used for scrolling to a # anchor in a page. However, it doesn't work with webrender right now anyway, so it's not technically a regression. https://github.com/servo/servo/issues/13736
</details>
Hrm, 

---

*[ports/cef/window.rs, line 298 at r2](https://reviewable.io:443/reviews/servo/servo/13711#-KTveqU9XxzQo4m4CsPD:-KTveqU9XxzQo4m4CsPE:b-rhejiz) ([raw file](https://github.com/servo/servo/blob/768a699fa5d822173b2247d6348413d09951a8bb/ports/cef/window.rs#L298)):*
> ```Rust
>     fn native_display(&self) -> NativeDisplay {
>         NativeDisplay::new()
>     }
> ```

@discomfitor Do we need these methods for the Linux CEF bindings?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/13711)*

**Comment by metajack** [comment]
Let's open bugs for the `-w` flag removal and the fixing of scrolling to anchors.

**Comment by pcwalton** [comment]
LGTM

**Comment by glennw** [comment]
@metajack https://github.com/servo/servo/issues/13736 https://github.com/servo/servo/issues/13761

**Comment by larsbergstrom** [comment]
@glennw Do you know if that last of an exposed binding will break CEF? Other than that and a rebase, it looks like this is g2g.

**Comment by glennw** [comment]
@larsbergstrom Rebased. I don't believe that native_display() removal affects the cef build, if that's what you're asking?

**Comment by emilio** [comment]
Yeah, that `native_display` method was only used to pass the appropriate X display to `rust-layers` AFAIK. In fact I think that now all the logic for that both in the CEF and glutin directories can be removed.

**Comment by glennw** [comment]
@larsbergstrom Rebased again, over the clip changes :)

**Comment by glennw** [comment]
@bors-servo retry

* #13360

**Comment by glennw** [comment]
@larsbergstrom Rebased and try passed, so should be good to go when you want to flick the switch :)

**Comment by emilio** [comment]
@bors-servo: p=1

 * Prone to bitrot

**Pull request merged**